app/services/charge_param_retriever.js

```diff
@@ -7,20 +7,25 @@ const cookie = require('../utils/cookies')
 
 exports.retrieve = req => {
   const chargeId = req.params.chargeId ? req.params.chargeId : req.body.chargeId
-  setLoggingField(req, PAYMENT_EXTERNAL_ID, chargeId)
 
+  setLoggingField(req, PAYMENT_EXTERNAL_ID, chargeId)
+  
   if (!chargeId) {
     logger.error('ChargeId was not found in request', getLoggingFields(req))
     return false
   } else if (!cookie.getSessionCookieName() || !cookie.isSessionPresent(req)) {
-    logger.info('Session cookie is not present', {
-      ...getLoggingFields(req),
-      referrer: req.get('Referrer'),
-      url: req.originalUrl,
-      method: req.method
-    })
-    return false
-  } else if (!cookie.getSessionVariable(req, 'ch_' + chargeId)) {
+    if(req.originalUrl && req.method.toLowerCase() === 'post' && req.originalUrl.includes(`/card_details/${chargeId}/3ds_required_in`)) {
```

Comment:
I'm concerned about the increasing complexity and side effects of this method. The pattern isn't established by this pull request but it may make it even harder to reason about in the future. 

The method is called `retrieve` but I think if we tried to encapsulate what it was doing it would be something along the lines of `safelyGetChargeExternalIdFromRequestOrFailIfCookieSessionIsInvalid` (exaggerated).

Adding this logic would make it `safelyGetChargeExternalIdFromRequestOrFailIfCookieSessionIsInvalidIfNotThe3DSRequiredInRoute`.

I wonder if we would be able to split the logic here so that requiring a valid cookie session would be a separate middleware from the one that gets the details of the charge from Connector. 

That then might look something like this from a very high level

```js 
const { requireValidCookieSession, fetchChargeDetails } = require('some/middleware')

// doesn't require a valid cookie session
app.post('/3ds/required/in', fetchChargeDetails, inHandler)

// does require a valid cookie session
app.post('/3ds/required/out', requireValidCookieSession, fetchChargeDetails, outHandler)
```

With this proposal if a particular route should require a session is declared at the top level and the logical flow can be followed. It's trying to avoid having to know the ins and outs of how all of the middleware work to understand which paths will require sessions and which will not (the fact that the middleware `retrieveCharge` will call a method `charge_param_retriever.retrieve` which will work one way for some routes and work another way for others for example).